2008-07-20  Ulrich Drepper  <drepper@redhat.com>

	[BZ #6712]
	* locale/setlocale.c (setlocale): Take the setlocale lock earlier.

--- libc/locale/setlocale.c	31 Mar 2008 00:37:03 -0000	1.69
+++ libc/locale/setlocale.c	20 Jul 2008 17:11:20 -0000	1.70
@@ -234,9 +234,16 @@ setlocale (int category, const char *loc
   if (locale == NULL)
     return (char *) _nl_global_locale.__names[category];
 
+  /* Protect global data.  */
+  __libc_rwlock_wrlock (__libc_setlocale_lock);
+
   if (strcmp (locale, _nl_global_locale.__names[category]) == 0)
-    /* Changing to the same thing.  */
-    return (char *) _nl_global_locale.__names[category];
+    {
+      /* Changing to the same thing.  */
+      __libc_rwlock_unlock (__libc_setlocale_lock);
+
+      return (char *) _nl_global_locale.__names[category];
+    }
 
   /* We perhaps really have to load some data.  So we determine the
      path in which to look for the data now.  The environment variable
@@ -250,12 +257,13 @@ setlocale (int category, const char *loc
   if (locpath_var != NULL && locpath_var[0] != '\0')
     {
       if (__argz_create_sep (locpath_var, ':',
-			     &locale_path, &locale_path_len) != 0)
-	return NULL;
-
-      if (__argz_add_sep (&locale_path, &locale_path_len,
-			  _nl_default_locale_path, ':') != 0)
-	return NULL;
+			     &locale_path, &locale_path_len) != 0
+	  || __argz_add_sep (&locale_path, &locale_path_len,
+			     _nl_default_locale_path, ':') != 0)
+	{
+	  __libc_rwlock_unlock (__libc_setlocale_lock);
+	  return NULL;
+	}
     }
 
   if (category == LC_ALL)
@@ -290,8 +298,13 @@ setlocale (int category, const char *loc
 		  break;
 
 	      if (cnt == __LC_LAST)
-		/* Bogus category name.  */
-		ERROR_RETURN;
+		{
+		error_return:
+		  __libc_rwlock_unlock (__libc_setlocale_lock);
+
+		  /* Bogus category name.  */
+		  ERROR_RETURN;
+		}
 
 	      /* Found the category this clause sets.  */
 	      newnames[cnt] = ++cp;
@@ -310,12 +323,9 @@ setlocale (int category, const char *loc
 	  for (cnt = 0; cnt < __LC_LAST; ++cnt)
 	    if (cnt != LC_ALL && newnames[cnt] == locale)
 	      /* The composite name did not specify all categories.  */
-	      ERROR_RETURN;
+	      goto error_return;
 	}
 
-      /* Protect global data.  */
-      __libc_rwlock_wrlock (__libc_setlocale_lock);
-
       /* Load the new data for each category.  */
       while (category-- > 0)
 	if (category != LC_ALL)
@@ -393,9 +403,6 @@ setlocale (int category, const char *loc
       struct locale_data *newdata = NULL;
       const char *newname[1] = { locale };
 
-      /* Protect global data.  */
-      __libc_rwlock_wrlock (__libc_setlocale_lock);
-
       if (CATEGORY_USED (category))
 	{
 	  /* Only actually load the data if anything will use it.  */
